### PR TITLE
16 handling connections requests

### LINKED ISSUE #16: Handling connections requests
### Problem Description
When a new client connects #6 to serversocket we need to handle that connection and wait for a CONNECT request.
Broker should then return CONNACK message.

We need inputStream and outputStream from client socket and after reading some data we can check if the incoming request is a CONNECT request.
https://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718028

`InputStream inputStream = socket.getInputStream();
 OutputStream outputStream = socket.getOutputStream();
 byte[] buffer = new byte[1024];

                int bytesRead = inputStream.read(buffer);
                if (bytesRead > 0 && buffer[0] == (byte) 0x10) {
                    System.out.println("Received MQTT CONNECT message from client");
                    byte[] connackMessage = new byte[]{(byte) 0x20, (byte) 0x02, (byte) 0x00, (byte) 0x00};
                    outputStream.write(connackMessage);
                    outputStream.flush();
                    System.out.println("Sent MQTT CONNACK message to client");`
                    
                  

- After a Network Connection is established by a Client to a Server, the first Packet sent from the Client to the Server MUST be a CONNECT Packet
- The Server MUST process a second CONNECT Packet sent from a Client as a protocol violation and disconnect the Client
- In a connect message the first byte is hexadecimal value 0x10

Response in form of CONNACK
https://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718033

To prevent denial of service attacks against our server we could add a limit time and disconnect the client if inputStream.read(buffer); hasn't returned in for example 30 seconds.
Java supports timeout on sockets read using the [setSoTimeout method.](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/net/Socket.html#setSoTimeout(int))

### PR DESCRIPTION
**Pull request by vonfredd**
---
Handling connection request for the first time

We should validate the connections made according to MQTT protocol

---

### Related Issue(s)
Closes #16 

### Proposed Changes
I suggest a separate class that can validate packages sent from a client to the server. The server can then respont to the client if connection is valid.

- **Change 1**: A new class called PackageReader.
- **Change 2**: Check connections in ClientHandler class before adding client to list.

Here are some example changes:
- Clienthandler get private final PackageReader field.
- We can read clients socket output and validate its package. If package is valid, the Connect message will be mapped with InetAdress of the clients socket. If a socket tries to send a second Connect package, the map inside PackageReader will trigger a statement that returns value false to Clienthandler, ending up closing the socket.

### CODE REVIEW

File: src/main/java/org/fungover/thunder/PackageReader.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        int bytesRead = inputStream.read(buffer);
+
+        if (isConnectPackage(bytesRead, buffer)) {
+            System.out.println("Received MQTT CONNECT message from client");
+            connectPackageSent.put(socket.getInetAddress(),true);
+            sendConnackToClient(outputStream);
+            System.out.println("Sent MQTT CONNACK message to client");
+        }
+
+        return false;
```
Comment (Win-ther) [inline src/main/java/org/fungover/thunder/PackageReader.java:34]: This needs to return true, if it returns false the socket will close and disconnect the client.

**Review by Win-ther — CHANGES_REQUESTED** [review]
isValidConnection method needs to return true. Other than that the code looks good.

File: src/main/java/org/fungover/thunder/PackageReader.java (hunk 1 above)
Comment (vonfredd) [reply src/main/java/org/fungover/thunder/PackageReader.java:34]: Correct, I did forget that. Thank you! 😄

**Review by Win-ther — DISMISSED** [review]

**Review by TrungMa1 — CHANGES_REQUESTED** [review]
Tests could be done for isConnectPackage method in PackageReader class.

File: src/main/java/org/fungover/thunder/PackageReader.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    private static void sendConnackToClient(OutputStream outputStream) throws IOException {
+        byte[] connackMessage = new byte[]{(byte) 0x20, (byte) 0x02, (byte) 0x00, (byte) 0x00};
+        outputStream.write(connackMessage);
+        outputStream.flush();
+    }
+
+    private boolean isConnectPackage(int bytesRead, byte[] buffer) {
```
[the remark this answers is not in the document]
Comment (vonfredd) [reply src/main/java/org/fungover/thunder/PackageReader.java:43]: Wouldn't it be a problem if two different clientHandler.handleConnections-threads call for the static method? They share the same method that belongs to the class so if one should return true and one should return false for example wouldn't it be possible to return both true?

**Comment by vonfredd** [comment]
> Tests could be done for isConnectPackage method in PackageReader class.
Absolutely , I will make some tests👍

File: src/main/java/org/fungover/thunder/PackageReader.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    private final Map<InetAddress, Boolean> connectPackageSent = new HashMap<>();
+
+    public boolean isValidConnection(Socket socket) throws IOException {
+        InetAddress client = socket.getInetAddress();
+
+        if (connectPackageSent.containsKey(client) && Boolean.TRUE.equals(connectPackageSent.get(client))) {
+            return false;
+        }
+
+        socket.setSoTimeout(30);
```
[the remark this answers is not in the document]
Comment (vonfredd) [reply src/main/java/org/fungover/thunder/PackageReader.java:21]: Thank you, I will fix that! 👍

File: src/main/java/org/fungover/thunder/PackageReader.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kappsegla) [reply src/main/java/org/fungover/thunder/PackageReader.java:43]: When only using parameters and local/method variables each thread will have its own set so no problems with that. Each thread has its own stack where these variables and return values are passed around.

[the remark this answers is not in the document]
Comment (vonfredd) [reply src/main/java/org/fungover/thunder/PackageReader.java:43]: Ok thanks. I change them to static 👍

**Review by TrungMa1 — APPROVED** [review]
Good job!

**Review by Win-ther — APPROVED** [review]

**Pull request merged**